pkg/controller/user-cluster-controller-manager/application-installation-controller/controller.go

```diff
@@ -112,18 +128,116 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+			return r.userClient.Delete(ctx, appInstallation)
+		} else {
+			return fmt.Errorf("applicationVersion: '%s' does not exist. can not install application", appInstallation.Spec.ApplicationRef.Version)
+		}
+	}
+
+	oldAppInstallation := appInstallation.DeepCopy()
+	appInstallation.Status.ApplicationVersion = appVersion
+
+	if err := r.userClient.Status().Patch(ctx, appInstallation, ctrlruntimeclient.MergeFrom(oldAppInstallation)); err != nil {
+		return fmt.Errorf("failed to update status with applicationVersion: %w", err)
+	}
+
+	// install application into user-cluster
+	if err := r.handleInstallation(ctx, log, appInstallation); err != nil {
+		return fmt.Errorf("handling installation of application installation: %w", err)
+	}
+
+	return nil
+}
+
+// getApplicationVersion finds the applicationVersion defined by appInstallation into the applicationDef and update the struct appVersion with it.
+// an error is returned if the applicationVersion is not found.
+func (r *reconciler) getApplicationVersion(appInstallation *appkubermaticv1.ApplicationInstallation, applicationDef *appkubermaticv1.ApplicationDefinition, appVersion *appkubermaticv1.ApplicationVersion) error {
+	desiredVersion := appInstallation.Spec.ApplicationRef.Version.String()
+	for _, version := range applicationDef.Spec.Versions {
+		if version.Version == desiredVersion {
+			version.DeepCopyInto(appVersion)
+			return nil
+		}
+	}
+	return fmt.Errorf("application version '%s' does not exist in applicationDefinition %s", desiredVersion, applicationDef.Name)
+}
+
+// handleInstallation install or update the application in the user cluster.
+func (r *reconciler) handleInstallation(ctx context.Context, log *zap.SugaredLogger, appInstallation *appkubermaticv1.ApplicationInstallation) error {
+	return r.appInstaller.Apply(ctx, log, r.seedClient, r.userClient, appInstallation)
+}
+
+// // handleDeletion uninstall the application in the user cluster.
+func (r *reconciler) handleDeletion(ctx context.Context, log *zap.SugaredLogger, appInstallation *appkubermaticv1.ApplicationInstallation) error {
+	if kuberneteshelper.HasFinalizer(appInstallation, appkubermaticv1.ApplicationInstallationCleanupFinalizer) {
+		if err := r.appInstaller.Delete(ctx, log, r.seedClient, r.userClient, appInstallation); err != nil {
+			return fmt.Errorf("failed to uninstall application: %w", err)
+		}
+
+		if err := kuberneteshelper.TryRemoveFinalizer(ctx, r.userClient, appInstallation, appkubermaticv1.ApplicationInstallationCleanupFinalizer); err != nil {
+			return fmt.Errorf("failed to remove application installation finalizer %s: %w", appInstallation.Name, err)
+		}
+	}
+
 	return nil
 }
 
+// traceWarning log the message in warning mode and raise a k8s event on appInstallation with the eventReason and the message.
+func (r *reconciler) traceWarning(appInstallation *appkubermaticv1.ApplicationInstallation, log *zap.SugaredLogger, eventReason, message string) {
+	log.Warnf(message)
```

Comment:
This can be _Warn()_

Reply:
thanks for catching that, i didn't complete my refactor :/